svm/src/account_loader.rs

```diff
@@ -95,6 +96,264 @@ pub struct FeesOnlyTransaction {
+        rollback_accounts: &RollbackAccounts,
+    ) {
+        let fee_payer_address = message.fee_payer();
+        match rollback_accounts {
+            RollbackAccounts::FeePayerOnly { fee_payer_account } => {
+                self.update_account(fee_payer_address, fee_payer_account);
+            }
+            RollbackAccounts::SameNonceAndFeePayer { nonce } => {
+                self.update_account(nonce.address(), nonce.account());
+            }
+            RollbackAccounts::SeparateNonceAndFeePayer {
+                nonce,
+                fee_payer_account,
+            } => {
+                self.update_account(nonce.address(), nonce.account());
+                self.update_account(fee_payer_address, fee_payer_account);
+            }
+        }
+    }
+
+    fn update_accounts_for_successful_tx(
+        &mut self,
+        message: &impl SVMMessage,
+        transaction_accounts: &[TransactionAccount],
+    ) {
+        for (i, (address, account)) in (0..message.account_keys().len()).zip(transaction_accounts) {
+            if !message.is_writable(i) {
+                continue;
+            }
+
+            // Accounts that are invoked and also not passed as an instruction
+            // account to a program don't need to be stored because it's assumed
+            // to be impossible for a committable transaction to modify an
+            // invoked account if said account isn't passed to some program.
+            if message.is_invoked(i) && !message.is_instruction_account(i) {
+                continue;
+            }
+
+            self.update_account(address, account);
+        }
+    }
+
+    fn update_account(&mut self, account_key: &Pubkey, account: &AccountSharedData) {
+        // If an account has been dropped, we must hide the data from any future transactions.
+        // We NEVER remove items from cache, or else we will fetch stale data from accounts-db.
+        // Because `update_account()` is only called on writable accounts using the same rules
+        // as commit, this (correctly) does not shadow rent-paying zero-lamport read-only accounts.
+        let account = if account.lamports() == 0 {
+            &AccountSharedData::default()
+        } else {
+            account
+        };
+
+        // Accounts are inspected during loading, and we only update writable accounts.
+        // Ergo, `inspected_as_writable` is always true.
+        self.account_cache
+            .entry(*account_key)
+            .and_modify(|cache_item| {
+                debug_assert!(cache_item.inspected_as_writable);
+                cache_item.account = account.clone();
+            })
+            .or_insert_with(|| AccountCacheItem {
+                account: account.clone(),
+                inspected_as_writable: true,
+            });
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub(crate) enum AccountUsagePattern {
+    Writable,
+    Instruction,
+    WritableInstruction,
```

Comment:
this seems functionally the same as `Writable`. is there somewhere that the distinction matters that I am missing?

Reply:
the two cases we care about are "writable vs read-only" and "writable or instruction vs neither," so writable vs writable instruction doesnt matter but writable instruction vs instruction does. i can cut this down to three variants tho, i was thinking too much about the two bools i was replacing (the reason i made this enum at all was to not have to worry about the order of two bools passed into functions)

Reply:
I think let's cut it down to 3, because otherwise it's less obvious from reading the code (see my comment below on fee-payer) that it is correct to mark it as `Writable` even though it may be used in an instruction.

Reply:
[824b350](https://github.com/anza-xyz/agave/pull/3146/commits/824b3502de2f0c0b43e101f279cafce73d739e74)